packages/common/src/utils/resolveDynamicValues.ts

```diff
@@ -0,0 +1,182 @@
+import { queryFeatures } from "@esri/arcgis-rest-feature-layer";
+import { getSelf, ISearchOptions, searchItems } from "@esri/arcgis-rest-portal";
+import { IStatisticDefinition } from "@esri/arcgis-rest-types";
+import { IArcGISContext } from "../ArcGISContext";
+import {
+  DynamicValueDefinition,
+  DynamicValues,
+  IDynamicItemQueryDefinition,
+  IDynamicPortalQueryDefinition,
+  IDynamicServiceQueryDefinition,
+} from "../core/types/DynamicValues";
+import { getProp } from "../objects/get-prop";
+import { setProp } from "../objects/set-prop";
+import { aggregateValues } from "./aggregateValues";
+import { memoize } from "./memoize";
+
+/**
+ * Create memoized versions of the searchItems and getSelf functions
+ */
+const memoizedItemSearch = memoize(searchItems);
+const memoizedPortalSelf = memoize(getSelf);
+
+/**
+ * Resolve a set of dynamic values. If the result is a DynamicValueDefinition, it will be resolved recursively
```

Comment:
Please describe the 'why' of the different source types (item-query, portal, service-query)